File: go/jsonrpc.go
package hyperserve

import (
	"encoding/json"
	"fmt"
	"log/slog"
)

// JSONRPCVersion is the JSON-RPC 2.0 version identifier
const JSONRPCVersion = "2.0"

// JSONRPCRequest represents a JSON-RPC 2.0 request message
type JSONRPCRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
	ID      interface{} `json:"id,omitempty"`
}

// JSONRPCResponse represents a JSON-RPC 2.0 response message
type JSONRPCResponse struct {
	JSONRPC string         `json:"jsonrpc"`
	Result  interface{}    `json:"result,omitempty"`
	Error   *JSONRPCError  `json:"error,omitempty"`
	ID      interface{}    `json:"id"`
}

// JSONRPCError represents a JSON-RPC 2.0 error object
type JSONRPCError struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

// Standard JSON-RPC error codes
const (
	ErrorCodeParseError     = -32700
	ErrorCodeInvalidRequest = -32600
	ErrorCodeMethodNotFound = -32601
	ErrorCodeInvalidParams  = -32602
	ErrorCodeInternalError  = -32603
)

// JSONRPCMethodHandler defines the signature for JSON-RPC method handlers
type JSONRPCMethodHandler func(params interface{}) (interface{}, error)

// JSONRPCEngine handles JSON-RPC 2.0 request processing
type JSONRPCEngine struct {
	methods map[string]JSONRPCMethodHandler
	logger  *slog.Logger
}

// NewJSONRPCEngine creates a new JSON-RPC engine
func NewJSONRPCEngine() *JSONRPCEngine {
	return &JSONRPCEngine{
		methods: make(map[string]JSONRPCMethodHandler),
		logger:  logger,
	}
}

// RegisterMethod registers a method handler with the JSON-RPC engine
func (engine *JSONRPCEngine) RegisterMethod(name string, handler JSONRPCMethodHandler) {
	engine.methods[name] = handler
	engine.logger.Debug("JSON-RPC method registered", "method", name)
}

// ProcessRequest processes a JSON-RPC request and returns a response
func (engine *JSONRPCEngine) ProcessRequest(requestData []byte) []byte {
	var request JSONRPCRequest
	
	// Parse the request
	if err := json.Unmarshal(requestData, &request); err != nil {
		engine.logger.Error("Failed to parse JSON-RPC request", "error", err)
		response := &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeParseError,
				Message: "Parse error",
				Data:    err.Error(),
			},
			ID: nil,
		}
		responseData, _ := json.Marshal(response)
		return responseData
	}
	
	// Validate JSON-RPC version
	if request.JSONRPC != JSONRPCVersion {
		engine.logger.Error("Invalid JSON-RPC version", "version", request.JSONRPC)
		response := &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeInvalidRequest,
				Message: "Invalid Request",
				Data:    fmt.Sprintf("expected jsonrpc version %s", JSONRPCVersion),
			},
			ID: request.ID,
		}
		responseData, _ := json.Marshal(response)
		return responseData
	}
	
	// Find method handler
	handler, exists := engine.methods[request.Method]
	if !exists {
		engine.logger.Error("JSON-RPC method not found", "method", request.Method)
		response := &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeMethodNotFound,
				Message: "Method not found",
				Data:    fmt.Sprintf("method '%s' not found", request.Method),
			},
			ID: request.ID,
		}
		responseData, _ := json.Marshal(response)
		return responseData
	}
	
	// Call method handler
	result, err := handler(request.Params)
	if err != nil {
		engine.logger.Error("JSON-RPC method execution error", "method", request.Method, "error", err)
		response := &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeInternalError,
				Message: "Internal error",
				Data:    err.Error(),
			},
			ID: request.ID,
		}
		responseData, _ := json.Marshal(response)
		return responseData
	}
	
	// Success response
	response := &JSONRPCResponse{
		JSONRPC: JSONRPCVersion,
		Result:  result,
		ID:      request.ID,
	}
	
	responseData, err := json.Marshal(response)
	if err != nil {
		engine.logger.Error("Failed to marshal JSON-RPC response", "error", err)
		errorResponse := &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeInternalError,
				Message: "Internal error",
				Data:    "failed to marshal response",
			},
			ID: request.ID,
		}
		errorResponseData, _ := json.Marshal(errorResponse)
		return errorResponseData
	}
	
	engine.logger.Debug("JSON-RPC request processed successfully", "method", request.Method)
	return responseData
}

// ProcessRequestDirect processes a JSON-RPC request object directly and returns a response object
func (engine *JSONRPCEngine) ProcessRequestDirect(request *JSONRPCRequest) *JSONRPCResponse {
	// Validate request
	if request.JSONRPC != JSONRPCVersion {
		engine.logger.Error("Invalid JSON-RPC version", "version", request.JSONRPC)
		return &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeInvalidRequest,
				Message: "Invalid Request",
				Data:    fmt.Sprintf("expected jsonrpc version %s", JSONRPCVersion),
			},
			ID: request.ID,
		}
	}
	
	// Find method handler
	handler, exists := engine.methods[request.Method]
	if !exists {
		engine.logger.Error("JSON-RPC method not found", "method", request.Method)
		return &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeMethodNotFound,
				Message: "Method not found",
				Data:    fmt.Sprintf("method '%s' not found", request.Method),
			},
			ID: request.ID,
		}
	}
	
	// Call method handler
	result, err := handler(request.Params)
	if err != nil {
		engine.logger.Error("JSON-RPC method execution error", "method", request.Method, "error", err)
		return &JSONRPCResponse{
			JSONRPC: JSONRPCVersion,
			Error: &JSONRPCError{
				Code:    ErrorCodeInternalError,
				Message: "Internal error",
				Data:    err.Error(),
			},
			ID: request.ID,
		}
	}
	
	// Build response
	response := &JSONRPCResponse{
		JSONRPC: JSONRPCVersion,
		Result:  result,
		ID:      request.ID,
	}
	
	engine.logger.Debug("JSON-RPC request processed successfully", "method", request.Method)
	return response
}

// GetRegisteredMethods returns a list of all registered method names
func (engine *JSONRPCEngine) GetRegisteredMethods() []string {
	methods := make([]string, 0, len(engine.methods))
	for method := range engine.methods {
		methods = append(methods, method)
	}
	return methods
}

File: rust/src/mcp/jsonrpc.rs
//! JSON-RPC 2.0 implementation for MCP

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use crate::json::{self, Value};

/// JSON-RPC version
pub const JSONRPC_VERSION: &str = "2.0";

/// JSON-RPC request
#[derive(Debug, Clone)]
pub struct JsonRpcRequest {
    /// JSON-RPC version
    pub jsonrpc: String,
    /// Method name
    pub method: String,
    /// Method parameters
    pub params: Option<Value>,
    /// Request ID
    pub id: Option<Value>,
}

impl JsonRpcRequest {
    /// Parse from JSON value
    pub fn from_json(value: Value) -> Result<Self, String> {
        let obj = value.as_object().ok_or("Request must be an object")?;
        
        let jsonrpc = obj.get("jsonrpc")
            .and_then(|v| v.as_str())
            .ok_or("Missing jsonrpc field")?
            .to_string();
            
        let method = obj.get("method")
            .and_then(|v| v.as_str())
            .ok_or("Missing method field")?
            .to_string();
            
        let params = obj.get("params").cloned();
        let id = obj.get("id").cloned();
        
        Ok(Self {
            jsonrpc,
            method,
            params,
            id,
        })
    }
    
    /// Convert to JSON value
    pub fn to_json(&self) -> Value {
        let mut obj = HashMap::new();
        obj.insert("jsonrpc".to_string(), Value::String(self.jsonrpc.clone()));
        obj.insert("method".to_string(), Value::String(self.method.clone()));
        
        if let Some(params) = &self.params {
            obj.insert("params".to_string(), params.clone());
        }
        
        if let Some(id) = &self.id {
            obj.insert("id".to_string(), id.clone());
        }
        
        Value::Object(obj)
    }
}

/// JSON-RPC response
#[derive(Debug, Clone)]
pub struct JsonRpcResponse {
    /// JSON-RPC version
    pub jsonrpc: String,
    /// Result (if success)
    pub result: Option<Value>,
    /// Error (if failed)
    pub error: Option<JsonRpcError>,
    /// Request ID
    pub id: Option<Value>,
}

/// JSON-RPC error
#[derive(Debug, Clone)]
pub struct JsonRpcError {
    /// Error code
    pub code: i32,
    /// Error message
    pub message: String,
    /// Additional error data
    pub data: Option<Value>,
}

impl JsonRpcResponse {
    /// Create a success response
    pub fn success(id: Option<Value>, result: Value) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            result: Some(result),
            error: None,
            id,
        }
    }
    
    /// Create an error response
    pub fn error(id: Option<Value>, error: JsonRpcError) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            result: None,
            error: Some(error),
            id,
        }
    }
    
    /// Convert to JSON value
    pub fn to_json(&self) -> Value {
        let mut obj = HashMap::new();
        obj.insert("jsonrpc".to_string(), Value::String(self.jsonrpc.clone()));
        
        if let Some(result) = &self.result {
            obj.insert("result".to_string(), result.clone());
        }
        
        if let Some(error) = &self.error {
            let mut error_obj = HashMap::new();
            error_obj.insert("code".to_string(), Value::Number(error.code as f64));
            error_obj.insert("message".to_string(), Value::String(error.message.clone()));
            if let Some(data) = &error.data {
                error_obj.insert("data".to_string(), data.clone());
            }
            obj.insert("error".to_string(), Value::Object(error_obj));
        }
        
        if let Some(id) = &self.id {
            obj.insert("id".to_string(), id.clone());
        } else {
            obj.insert("id".to_string(), Value::Null);
        }
        
        Value::Object(obj)
    }
}

/// Type alias for method handler
pub type MethodHandler = Arc<dyn Fn(Option<Value>) -> Result<Value, JsonRpcError> + Send + Sync>;

/// JSON-RPC engine for processing requests
pub struct JsonRpcEngine {
    methods: Arc<Mutex<HashMap<String, MethodHandler>>>,
}

impl JsonRpcEngine {
    /// Create a new JSON-RPC engine
    pub fn new() -> Self {
        Self {
            methods: Arc::new(Mutex::new(HashMap::new())),
        }
    }
    
    /// Register a method handler
    pub fn register_method<F>(&self, name: &str, handler: F)
    where
        F: Fn(Option<Value>) -> Result<Value, JsonRpcError> + Send + Sync + 'static,
    {
        let mut methods = self.methods.lock().unwrap();
        methods.insert(name.to_string(), Arc::new(handler));
    }
    
    /// Process a JSON-RPC request
    pub fn process_request(&self, request: &JsonRpcRequest) -> JsonRpcResponse {
        // Validate JSON-RPC version
        if request.jsonrpc != JSONRPC_VERSION {
            return JsonRpcResponse::error(
                request.id.clone(),
                JsonRpcError {
                    code: -32600,
                    message: "Invalid Request".to_string(),
                    data: Some(Value::String(format!("Expected jsonrpc version {}", JSONRPC_VERSION))),
                },
            );
        }
        
        // Find method handler
        let methods = self.methods.lock().unwrap();
        let handler = match methods.get(&request.method) {
            Some(h) => Arc::clone(h),
            None => {
                return JsonRpcResponse::error(
                    request.id.clone(),
                    JsonRpcError {
                        code: -32601,
                        message: "Method not found".to_string(),
                        data: Some(Value::String(format!("Method '{}' not found", request.method))),
                    },
                );
            }
        };
        
        // Drop the lock before calling the handler
        drop(methods);
        
        // Call method handler
        match handler(request.params.clone()) {
            Ok(result) => JsonRpcResponse::success(request.id.clone(), result),
            Err(error) => JsonRpcResponse::error(request.id.clone(), error),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_jsonrpc_request_parsing() {
        let json_str = r#"{
            "jsonrpc": "2.0",
            "method": "test",
            "params": {"foo": "bar"},
            "id": 1
        }"#;
        
        let value = json::parse(json_str).unwrap();
        let request = JsonRpcRequest::from_json(value).unwrap();
        
        assert_eq!(request.jsonrpc, "2.0");
        assert_eq!(request.method, "test");
        assert!(request.id.is_some());
    }
    
    #[test]
    fn test_jsonrpc_engine() {
        let engine = JsonRpcEngine::new();
        
        // Register a test method
        engine.register_method("echo", |params| {
            Ok(params.unwrap_or(Value::Null))
        });
        
        // Test successful call
        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            method: "echo".to_string(),
            params: Some(Value::String("hello".to_string())),
            id: Some(Value::Number(1.0)),
        };
        
        let response = engine.process_request(&request);
        assert_eq!(response.result, Some(Value::String("hello".to_string())));
        assert!(response.error.is_none());
        
        // Test method not found
        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            method: "unknown".to_string(),
            params: None,
            id: Some(Value::Number(2.0)),
        };
        
        let response = engine.process_request(&request);
        assert!(response.result.is_none());
        assert!(response.error.is_some());
        assert_eq!(response.error.unwrap().code, -32601);
    }
}